streams-runtimes/streams-runtime-threaded/src/main/java/org/apache/streams/threaded/controller/ThreadingController.java

```diff
@@ -0,0 +1,253 @@
+
+import com.google.common.util.concurrent.*;
+import org.apache.commons.lang3.concurrent.BasicThreadFactory;
+import org.slf4j.Logger;
+
+import java.util.Date;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Condition;
+import org.slf4j.LoggerFactory;
+
+public class ThreadingController {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ThreadingController.class);
+
+    private final String name;
+    private final int maxNumberOfThreads;
+    private final int priority;
+    private final ThreadPoolExecutor threadPoolExecutor;
+    private final ListeningExecutorService listeningExecutorService;
+    private final Condition lock = new SimpleCondition();
+    private final AtomicInteger numThreads;
+    private final AtomicLong lastWorked = new AtomicLong(new Date().getTime());
+    private final AtomicLong numberOfObservations = new AtomicLong(0);
+    private final AtomicDouble sumOfObservations = new AtomicDouble(0);
+    private final AtomicInteger workingNow = new AtomicInteger(0);
+
+    private volatile double lastCPUObservation = 0.0;
+
+    private static final long SCALE_CHECK = 2000;
+    private Double scaleThreshold = .85;
+    private ThreadingControllerCPUObserver threadingControllerCPUObserver = new DefaultThreadingControllerCPUObserver();
+
+    private static final Integer NUM_PROCESSORS = Runtime.getRuntime().availableProcessors();
+
+    public static final ThreadingController INSTANCE_LOW_PRIORITY = new ThreadingController("Apache Streams [low]", NUM_PROCESSORS, NUM_PROCESSORS * 2, Thread.NORM_PRIORITY - 2);
+    public static final ThreadingController INSTANCE = new ThreadingController("Apache Streams [default]", NUM_PROCESSORS, NUM_PROCESSORS * 5, Thread.NORM_PRIORITY);
+    public static final ThreadingController INSTANCE_HIGH_PRIORITY = new ThreadingController("Apache Streams [high]", NUM_PROCESSORS, NUM_PROCESSORS * 7, Thread.NORM_PRIORITY + 2);
+
+    /**
+     * Use for very low priority items... The thread-pool that runs this runs at priority
+     * (Thread.NORM_PRIORITY - 2) = 3
+     * @return
+     * The threading controller
+     */
+    public static ThreadingController getInstanceLowPriority() {
+        return INSTANCE_LOW_PRIORITY;
+    }
+
+    public static ThreadingController getInstance() {
+        return INSTANCE;
+    }
+
+    public static ThreadingController getInstanceHighPriority() {
+        return INSTANCE_HIGH_PRIORITY;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * The current number of threads in the core pool
+     * @return
+     * Integer representing the number of threads in the core pool.
+     */
+    public Integer getNumThreads()              { return this.numThreads.get(); }
```

Comment:
Please check the formatting in this class.  These are not the conventions used elsewhere in this project.